### PR TITLE
[feature] FSDP: enable pytorch SyncBN

### PR DESCRIPTION
**Pull request by min-xu-ai**
- not fully validated yet but at least not asserting
- this enables VISSL to move forward with its next PR

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/utils/parallel.py [hunk 1]
```diff
@@ -43,3 +43,15 @@ def validate_process_group(device: torch.device, process_group: ProcessGroup) ->
...
+def enable_pytorch_sync_bn(module: torch.nn.Module) -> None:
+    """Call _specify_ddp_gpu_num for all pytorch SyncBN layers so that it
+       is happily running even without DDP. E.g. this is used by FSDP.
+    """
+    for layer in module.modules():
+        if isinstance(layer, torch.nn.modules.SyncBatchNorm):
+            # Number "1" below meant to be the number of GPUs for each DDP worker.
+            # (i.e. "device_ids" in DDP. As far as I see, the value is not actually
+            # used, but this call needs to be made to avoid an exception.
+            layer._specify_ddp_gpu_num(1)  # type: ignore
```
Comment (myleott) [inline fairscale/utils/parallel.py:57]: Wow, I didn't realize this was being done in upstream PyTorch DDP, nice find 😄 

For reference: https://github.com/pytorch/pytorch/blob/e09e97ebf9946aa9b9582b27811bb7a9c3de392f/torch/nn/parallel/distributed.py#L1284-L1290

File: tests/nn/data_parallel/test_fsdp_regnet.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            self.bn = FSDP(
+                BatchNorm2d(2),
+                mixed_precision=False,
+                process_group=dist.new_group(ranks=[rank]),
+                flatten_parameters=False,
+            )
+
+        def forward(self, x):
+            x = self.conv(x)
+            x = self.bn(x)
```
Comment (myleott) [inline tests/nn/data_parallel/test_fsdp_regnet.py:48]: is the output type here always float32? Perhaps add:
```python
assert x.dtype == torch.float32
```

**Review by myleott — APPROVED** [review]
Awesome!

File: tests/nn/data_parallel/test_fsdp_regnet.py (hunk 2 above)
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_regnet.py:48]: It is a very good suggestion to add that. The result is something I wasn't expecting. The input to this "FP32 BN" can be either fp16 or fp32 and the output always match the input ind type. The internal weights of BN is kept at FP32. Let me check what kind of precision is in fact required in vision.

Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_regnet.py:48]: I also verified that autocast doesn't change this behavior for BN. It seems that we need to explicitly manage it for BN.

Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_regnet.py:48]: OK, I found out the answer. I think we only keep the weights fp32 and inputs are still fp16, base on the description here:

image

from this page: https://nvidia.github.io/apex/amp.html

**Pull request merged**